core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -158,6 +159,8 @@
 
 public class ExpressionInterpreter
 {
+    private static final Map<OptimizationKey, Object> optimizationCache = new WeakHashMap<>();
```

Comment:
`OptimizationKey` does work with `WeakHashMap` since no one holds to `OptimizationKey`. Your key must be `Expression`